### PR TITLE
DEP: deprecate asscalar

### LINKED ISSUE #4701: np.asscalar should pass through scalars
**Suggestion**
It would be very convenient if `np.asscalar()` would pass through scalars.  The use-case is to ensure that an argument is a scalar, regardless what is passed in (as long as it behaves as a scalar).

``` python
>>> np.asarray(1.0)
1.0
>>> np.asarray(1j)
1j
```

Presently these fail as the implementation simply looks for the `.item()` method, which scalars do not possess:

`````` python
>>> np.asarray(1.0)
AttributeError                            Traceback (most recent call last)
<ipython-input-13-c6bc08be135e> in <module>()
----> 1 np.asscalar(1.0)

.../python2.7/site-packages/numpy/lib/type_check.pyc in asscalar(a)
    461 
    462     """
--> 463     return a.item()
    464 
    465 #-----------------------------------------------------------------------------

AttributeError: 'float' object has no attribute 'item'```
``````

**Rational**
It is common to use `np.asarray()` to ensure that the argument is an array (in the case that it might be a list etc.).  In an analogous use case, it would be useful to be able to use `np.asscalar()` to ensure that an argument is a scalar (as long as it can be converted to a scalar).

**Workaround**
Convert the argument to an array first.

``` python
np.asscalar(np.asarray(x))
```

**Resolution**
Perhaps code like the following would work:

``` python
def asscalar(a):
    try:
        return a.item()
    except AttributeError, e:
        return np.asarray(a).item()
```

### PR DESCRIPTION
**Pull request by mattip**
Closes #4701 (issue). Closes #11856 (PR). Closes #10256 (PR). Closes #10659 (PR).

This issue was labelled "easy" which led to the multiple PRs to fix it. I propose to deprecate it instead, as it is neither used nor tested in numpy, and is a thin wrapper to [`ndarray.item()`](http://www.numpy.org/devdocs/reference/generated/numpy.ndarray.item.html) but accepts none of the `args`.

SciPy does not use `asscalar`, astropy has two uses of `asscalar` which would need fixing.

### CODE REVIEW

**Comment by shoyer** [comment]
:+1: This seems like a nice resolution to me. The function is basically useless.

**Comment by charris** [comment]
Hmm, what is the reason to only deprecate it in the docstring? Should it raise a deprecation warning?

**Comment by rgommers** [comment]
Agree that it needs a `DeprecationWarning`. Did a quick search - zero usage in `scipy`, `scikit-image` and `statsmodels`, 1 occurrence in `scikit-learn`. Won't impact users much.

**Comment by mattip** [comment]
@charris thanks. Added actual deprecation warning.

File: numpy/lib/type_check.py [hunk 1]
```diff
@@ -467,7 +467,10 @@ def real_if_close(a,tol=100):
 
 def asscalar(a):
     """
-    Convert an array of size 1 to its scalar equivalent.
+    Convert an array of size 1 to its scalar equivalent. Deprecated, use
+    `numpy.ndarray.item()` instead.
```
Comment (eric-wieser) [inline numpy/lib/type_check.py]: This second sentence should be indented under the `.. deprecated` block

File: numpy/lib/type_check.py [hunk 2]
```diff
@@ -486,6 +489,10 @@ def asscalar(a):
     24
 
     """
+
+    # 2018-10-10, 1.16
+    warnings.warn('asscalar will be removed in v1.18 of numpy',
+                  VisibleDeprecationWarning)
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/lib/type_check.py]: fixed

File: numpy/lib/type_check.py (hunk 1 above)
Comment (mattip) [reply numpy/lib/type_check.py]: moved

File: numpy/lib/type_check.py [hunk 3]
```diff
@@ -486,6 +491,10 @@ def asscalar(a):
     24
 
     """
+
+    # 2018-10-10, 1.16
+    warnings.warn('np.asscalar(a) will be removed in v1.18 of numpy, use '
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/lib/type_check.py]: Thanks for fixing

**Comment by charris** [comment]
Thanks Matti.

**Comment by the-admax** [comment]
Sorry, but why this handy function got deprecated in favor of `x.item()`? I find the functional approach more robust in dynamic typing world and providing more flexibility than attribute access when type of `x` is not known. For instance, this is a common case to use python `print` or string formatting functions when a value coming from array operation is actually a deeply nested scalar like `[[1]]`. So what is a proper alternative for this deprecated function?

**Comment by eric-wieser** [comment]
@the-admax: `ascalar` was implemented as `x.item()` anyway, so both would fail if `x` was not a numpy type.

**Pull request merged**